### ISSUE TITLE
False negative for mathcomp's gcdn function

### ISSUE THREAD

**Issue by anton-trunov** [QuickChick/QuickChick#201]
QuickChick reports that `gcdn` function is not commutative (which it is):
```coq
From mathcomp Require Import ssreflect ssrnat div.
From QuickChick Require Import QuickChick.

QuickChick
   (fun (s : nat) (t : nat) =>
      eqn
        (gcdn s t)
        (gcdn t s)).
(*
0
1
*** Failed after 2 tests and 1 shrinks. (0 discards)
 *)
```

Running the "counter-example" inside Coq returns `true`:
```coq
Compute
   (fun (s : nat) (t : nat) =>
      eqn
        (gcdn s t)
        (gcdn t s)) 0 1.
     (* = true *)
     (* : bool *)
```

I might be wrong but my guess is that happens because `nat` and operations on it extracted as OCaml's `int` and the corresponding (exception-throwing non-total) operations like `mod`.

**Comment by lemonidas** [comment]
You're right, it's the extraction unsoundness that yields this "counter-example". 

I'm not sure what the best way to address this moving forward is (so @Lysxia @liyishuai I'd like your thoughts!):
- One option would be to have the default be to extract all nats to nats, write users to import an "Extract Nat to Ocaml Int" module whenever they want performance over soundness in corner cases like the one here. This would significantly impact performance in the "normal" case however, as we perform a lot of computations on large numbers (for sizes, fuel, etc.).
- Another option would be almost the same, but have our own type of numbers that we use for internal quickChick bookkeeping and that we always extract to ocaml's int. If the user's coq code is nat-computation heavy it will still be slow as a default, but it might be a good balance.
- Alternatively, if we had another "Extract Nat Pedanticaly" module we could switch the default behavior of the previous option, but I'm not sure how to accomplish that. Maybe a plugin that purges extraction entries?

**Comment by llelf** [comment]
@anton-trunov oops :) that's bug in my code. #202

**Comment by liyishuai** [comment]
In the short term, #202 should solve this specific issue.
We should indeed allow users to configure how to extract numbers (`int` for better performance, `big_int` for #102, _etc_).

**Comment by jmgrosen** [comment]
I'm running into this because I want to use stdpp's gmap, but it uses big numbers everywhere, and so totally breaks when extracted under this regime... Being able to configure how to extract numbers would be great!
